### PR TITLE
Admin access to partition

### LINKED ISSUE #7631: Refactor BrokerAdminServiceImpl to use privileged access to partition
**Description**

Currently the `BrokerAdminServiceImpl` takes a list of Zeebe partitions in its constructor. This should be decoupled.

* Introduce new interface `PartitionAdminAcess` (see https://github.com/camunda-cloud/zeebe/blob/0445948961121032c850d3509d41610497fc0600/broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionAdminAccess.java for draft) (for clarification: this is one interface to admin all local partitions, it's not a per partition interface; also the futures returned should be `ActorFutrue` instead)
* Make `BrokerAdminServiceImpl` use this
* Implement this interface
* Make `PartitionManagerImpl` create an instance of this interface
* Make `Broker` pass the admin access interface to `BrokerAdminService`
* Remove `pause/ResumeProcessing` and related methods from public interface of `PartitionContext` and maybe even `ZeebePartition` and find a less prominent way to make this functionality available

**Constraints**
The admin interface, or the getter to get an instance of that interface shall not be exposed via the `PartitionManager` interface. The only way to get hold of the admin interface is through the implementation class, which is only known in `Broker`

### PR DESCRIPTION
**Pull request by pihme**
## Description

The admin service uses functionality to directly intervene in the ordinary operation of a partition. Among others, it allows to pause and resume processing or exporting. Since this behavior has a high impact, precautions needed to be taken that it is not used unintentionally. Previously, this behavior was directly available in `ZeebePartition`. 

The PR implements the following changes:
- introduce `PartitionAdminAccess` interface to capture the admin behavior of a partition
- move corresponding logic from `ZeebePartition` and `PartitionStartupAndTransitionContextImpl` into `ZeebePartitionAdminAccess` and `PartitionAdminControlImpl`
- change `BrokerAdminServiceImpl` to use the admin access interface
- introduce a stream collector to aggregate different `ActorFuture` objects into a single future

One thing that was not done in order to limit the scope: refactoring `ZeebePartitionAccess` and `PartitionAdminControlImpl` to either a single class or to two classes with a clearer separation of concerns. My general sense was that there is a reason for the current separation of concerns, and I didn't want to mess with it for now.

There are probably also other areas where refactoring could simplify code.

## Related issues

closes #7631

## Definition of Done

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [X] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:20]: :wrench: Adding this here since it applies to most methods, but I would not log the errors in this class - we return a future, and the caller should take care of handling the error (including logging). Otherwise you get strange double logging, and often are missing context. We should only log at the source if there's additional context we want to add (and in that case it would not be ERROR level). At the same time, I understand this is only code you moved, so I'm not going to insist at this moment.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManagerImpl.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManagerImpl.java:113]: :upside_down_face: Can we extract this to a temporary variable? I find it much more readable when debugging, since you immediately have the watch. This is really minor though, feel free to ignore.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/NoOpPartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/NoOpPartitionAdminAccess.java:50]: :question: Why warning? Is this because we expect this to never happen, or only at the beginning?

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:35]:
```suggestion
    final ActorFuture<Void> completed = concurrencyControl.createFuture();
```

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:52]:
```suggestion
    final ActorFuture<Void> completed = concurrencyControl.createFuture();
```

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:73]:
```suggestion
    final ActorFuture<Void> completed = concurrencyControl.createFuture();
```

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:93]:
```suggestion
    final ActorFuture<Void> completed = concurrencyControl.createFuture();
```

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:114]:
```suggestion
    final ActorFuture<Void> completed = concurrencyControl.createFuture();
```

**Review by npepinpe — CHANGES_REQUESTED** [review]
:rocket:

Some notes:

- :wrench: Generally I think the whole actor future collector/aggregation can be simplified. The accumulator's job is simply to merge two list of futures together, so it could be replaced by literally just using `List<ActorFuture<V>>`. The combiner is just returning a new list (or expanding one of them) which contains all futures, so again we can just use `List<ActorFuture<V>>`. And the finisher can then just create a new `CompletionWaiter` which does the final computation, and return this. I don't think we need the `ActorFutureCollector` at all - though I would suggest renaming the `ActorFutureAggregator` to `ActorFutureCollector`, since it's meant to be used as a `Collector` :smile:  Something like this (but not as ugly :smile:):

```java
  @Override
  public Supplier<List<ActorFuture<V>>> supplier() {
    return ArrayList::new;
  }

  @Override
  public BiConsumer<List<ActorFuture<V>>, ActorFuture<V>> accumulator() {
    return List::add;
  }

  @Override
  public BinaryOperator<List<ActorFuture<V>>> combiner() {
    return (listA, listB) -> {
      listA.addAll(listB);
      return listA;
    };
  }

  @Override
  public Function<List<ActorFuture<V>>, ActorFuture<List<V>>> finisher() {
    return futures -> new CompletionWaiter<>(concurrencyControl, futures).get();
  }
```

- Regarding the interface you created and my comments for documentation, I understand it's frustrating since you didn't actually add this code originally so you may not necessarily know the undocumented contract :smile: I'm happy to help out there, or have these be follow up issues so we can merge this, but I'm also afraid if I don't ask now nobody will ever document them.
- :+1: Happy to see things removed from the `PartitionContext` and `ZeebePartition` :smile: 
- I understand some suggestions were about code you moved, but if we don't update the code now, I'm not sure when would be the right time. Let me know what you think here.
- :wrench:  In my opinion, we could have split the PR in two at least, with the actor stuff in one, and the rest in another. It took me more time to understand what the actor collector was doing, but that's also because I'm not super familiar with that. No need to split it now though, just mentioning. Happy to hear if you disagree :smile:

File: broker/pom.xml
Comment (pihme) [reply broker/pom.xml:121]: :warning:  This would introduce a circular dependency between `test-util` and `util`

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java:22]: This (and other comments): Let's just say this was also my first implementation of a collector and there were different iterations of it. And I didn't spot the simplification you mentioned earlier. So I will use this one, and maybe not answer each and every comment

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java:22]: Expanded the javadoc. Not sure what you want to see in the example. Please suggest what would be helpful here.

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java:61]: Don't know it myself, tbh. My guess is that it would sequence the call to the collector, but not to any other calls as part of the parallel stream. But it's just a guess though,

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java:44]: I am pretty sure it's called exactly once.

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java:50]: Not really for performance. The main reason is random access: ` `results[currentIndex] = error == null ? Either.right(result) : Either.left(error);`

I did have a version in which I initialized a list with a bunch of nulls, but that felt weird, and then I remembered arrays.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionAdminAccess.java
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionAdminAccess.java:12]: Happy to receive help on this one, or defer it to a follow up issue.

a) I have no idea what the contract is
b) I don't even know what a contract can look like in an Actor world, because basically anything I can guarantee is that I will send a message to an actor. What the actor does with the message is then an implementation detail which I cannot really guarantee. But maybe I am overcomplicating things here.

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionAdminControl.java
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionAdminControl.java:14]: same here

File: util/src/main/java/io/camunda/zeebe/util/sched/Actor.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/Actor.java:17]: I had not thought about it. And now thinking about it I am undecided.
On the one hand I would like it better if it had a method `getConcurrencyControl` - intuitively.

On the other hand, I would probably return the `ActorControl` and then somebody could start to cast it :thinking: 
Also, there is Demeter's law to consider.

My main problem is that neither feels more natural to me. Mostly, because the actor class and the actor control class both feel unnatural to me. So I don't know whether I see the actor as an entity that I use for scheduling, or as an entity that exposes business logic.

Happy to hear your thoughts on this one.

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionAdminControlImpl.java
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/PartitionAdminControlImpl.java:24]: I want to get to a point where a `PartitionContext` is immutable, but can be swapped (in `ZeebePartition`) at runtime, e.g. after a transition. 

The `PartitionAdminControlImpl` shall only be created once. That's why it gets lookup methods instead of instances for its dependencies. Whether the lookup is for four individual things or one special interface is secondary. I wouldn't want to use `PartitionContext` because at least now it is both two big and missing some of the stuff we need.

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java:20]: I would leave it for now.

Logging at the source also has the advantage that you can be sure it is logged, regardless of where it is called from. Yes, double logging is inconvenient, but less so than forgetting somewhere to log at all. Also, this is not a high frequency call site. Also, in this instance, the call sites do not log the errors.

Maybe another addition to the style guide? Seems you have some vision of where logging should happen.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/NoOpPartitionAdminAccess.java
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/NoOpPartitionAdminAccess.java:50]: The way the class is used now, it would only happen during startup. But I don't know where this class will be used in the future. So the log message is to indicate that it's probably a mistake when this is called.

However, happy to reduce log level or remove logging altogether. Please advise.

File: broker/pom.xml
Comment (npepinpe) [reply broker/pom.xml:121]: Ah, the utility is also used in the `util` tests? Let's go with it now, but this is another strong indication that our modules are not well defined. I guess I need to think a bit about what we want to do here.

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java
Comment (npepinpe) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java]: Thanks, was just curious :slightly_smiling_face:

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/NoOpPartitionAdminAccess.java
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/NoOpPartitionAdminAccess.java:50]: Warning level indicates to users that something may be wrong, but can resolve itself without any intervention, and would only require manual intervention if it persists. If we never expected it to happen for example, then a warning is too low, and if we expected it to happen at sporadic intervals, then it would be too high.

If I understand correctly, we expect it to happen at the moment, but only at startup? So it's effectively indicating that something is not working, but we expect it to recover itself. Warning is fine then.

File: util/src/main/java/io/camunda/zeebe/util/sched/Actor.java
Comment (npepinpe) [reply util/src/main/java/io/camunda/zeebe/util/sched/Actor.java:17]: Demeter's Law is a good point. Let's leave it as is for now.

:thought_balloon: My own thoughts are that we should move to a more natural (at least, if you've worked with the actor model before :smile:) way of defining behaviors, and spawning actors which will run these. So the actor isn't the Broker, or the PartitionManager - the actor is simply the message inbox and dispatching mechanism, and you set the behavior (i.e. the business logic, e.g. Broker, PartitionManager) of handling these messages one at a time. In that sense, the Actor would be the concurrency control. But as it is now, these things are blurred, and would take some effort to decouple.

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFuture.java
Comment (npepinpe) [inline util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFuture.java:59]: :x: I think we also defined it on `CompletableActorFuture`, so we could now remove this and that `CompletableActorFuture` also implements `BiConsumer`.

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java
Comment (npepinpe) [inline util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java:73]: :question: I'm still confused by the characteristics :smile: Why did we add CONCURRENT now?

**Review by npepinpe — APPROVED** [review]
:+1: 

There is one small thing, we can remove the `BiConsumer<V, Throwable>` from `CompletableActorFuture` now I think. But nothing that needs another review

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java
Comment (npepinpe) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureAggregator.java]: The expanded one is more than enough now :+1:

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionAdminAccess.java
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionAdminAccess.java:12]: It's not about `PartitionAdminAccess` as an actor, but rather - what is it used for? What are its responsibilities? If I call `pauseExporting`, what happens? We know what happens - if the operation is successful, exporting is stopped, and this is persisted across restarts, such that no exporting will take place for this partition on this node until `resumeExporting` is called. It's also safe to call multiple times.

I'm happy to write this in a follow up PR though, especially since here you just extracted this, but in general I would like that we document public interfaces.

File: util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java
Comment (pihme) [reply util/src/main/java/io/camunda/zeebe/util/sched/future/ActorFutureCollector.java:73]: Actually, this is wrong. Yesterday I didn't think hard about it. Today I thought I could set this characteristic because I thought it is about the combiner function, which works concurrently. The JavaDoc isn't too detailed but it references the accumulator, which does not work concurrently. So I removed it again.

**Comment by pihme** [comment]
bors retry

**Pull request merged**
